src/librustc_lint/types.rs

```diff
@@ -615,6 +615,12 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
             return FfiSafe;
         }
 
+        // `extern "C" fn` functions can have type parameters, which are not FFI-safe,
+        // but they are currently ignored for the purposes of this lint, see #65134.
+        if ty.has_param_types() {
```

Comment:
What if you remove this and instead add a case for `ty::Param` below?
Then it would still warn for `Vec<T>` but not, say, `*mut T` or `MyCStruct<T>`.

Reply:
And same for `ty::Projection` - right now `<T as Iterator>::Item`, or similar, would make the lint ICE.

Reply:
I changed it to the way it is now because checking for `ty::Param` only would ICE on `<T as Iterator>::Item` (not exactly that, but a projection, in some other test).

Reply:
Because of this line: 
 https://github.com/davidtwco/rust/blob/8b2d8e19ec74f55af2cfa31bc642c5927c43fdeb/src/librustc_lint/types.rs#L889

Reply:
Oh, I see. If I make both `ty::Param` and `ty::Projection` FFI-safe then I can handle more cases.

Reply:
Yes, and I'm saying you should allow both.

Reply:
Fixed, thanks!